staging/src/k8s.io/client-go/tools/watch/until.go

```diff
@@ -95,6 +95,17 @@ func UntilWithoutRetry(ctx context.Context, watcher watch.Interface, conditions
 	return lastEvent, nil
 }
 
+// Until wraps the watchFunc with RetryWatcher making sure that watcher gets restarted in case of errors.
+// The initialResourceVersion will be given to watchFunc when first called.
+// Remaining behaviour is identical to function UntilWithoutRetry. (See above.)
+// Until can deal with API timeout or lost connections.
+// It guarantees you to see all events and in the order they happened.
+// Due to this guarantee there is no way it can deal with 'Resource version too old error'. It will fail in this case.
+// The most frequent usage would be a test where you want to verify exact order of events.
+func Until(ctx context.Context, initialResourceVersion string, watcherFunc WatcherFunc, conditions ...ConditionFunc) (*watch.Event, error) {
```

Comment:
verify `initialResourceVersion` is not "" or "0" (return error if it is)?

Reply:
could you help me understand why those are invalid values in this context?

Reply:
The synthetic ADDED watch events you get back for existing objects when starting a watch without specifying a resourceVersion are not ordered by resourceVersion, so restarting a broken watch request based on the resourceVersion of one of the received events (as the RetryWatcher does) is not safe.

That could be improved in the future, and if it is, the restriction here could be lifted.

Reply:
watch cache?
https://github.com/kubernetes/kubernetes/blob/2cbb16bc8dd456c5db72c1667926abdbc87c32c7/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L434-L461
listing directly from the store and breaking the event order sounds like a bug to me. I'll send a PR to fix it and we don't have to hack it here, also I want this to work from `""`.

Reply:
It is not watch cache specific. It happens on resources that do not use watch cache as well

Reply:
hm, if we don't use watch cache there, we are reading etcd, which I suppose has index on RV so reading it from the index should be preferable and ordered. (In the worst case the watch start 0 is basically a list there which we could order before sending it.)

What other places of breaking the invariant except watch cache init are you aware? are those just regular etcd reads? I think we need to fix those. The invariant of a watch call is (should be) that events are deliver in the order they happened (ordered by RV).

Reply:
> which I suppose has index on RV

That gets compacted after 5 minutes. List from etcd is not ordered by RV.

> In the worst case the watch start 0 is basically a list there which we could order before sending it

that is potentially a sort of a list of tens or hundreds of thousands of items

Reply:
https://github.com/kubernetes/kubernetes/issues/13969

And I agree with Jordan it's not easily fixable.

Reply:
requiring a specific resource version to start with here doesn't seem unreasonable to me, especially since this method's purpose is to ensure observation of every watch event, a guarantee you don't get when starting from ""/"0"

Reply:
I consider correctness our decision point. 

If my watch call get interrupted in the initial list phase, which I have no way to detect because it's just a first part of that continuous stream - how do I restart the watch? if I start from last RV received and the stream is not ordered I'll receive some events twice or skip some.

Reply:
I admit this may be slipping to how a single WATCH call is broken, not this handling.

I don't think we need v2 to fix that, although it would avoid the issue we are having but v2 premise is API change which gets hard.

I am pretty new to this part of the machinery, but looks like the fix so far requires a ´few lines in a watch cache and listing from etcd using WithSort()

Reply:
> If my watch call get interrupted in the initial list phase, which I have no way to detect because it's just a first part of that continuous stream - how do I restart the watch? if I start from last RV received and the stream is not ordered I'll receive some events twice or skip some.

Yes, that is currently the issue with watches started from "" or "0".

> I admit this may be slipping to how a single WATCH call is broken, not this handling.

I'm open to investigating improving that, but in the meantime, this should prevent usage we know is flawed.

> I am pretty new to this part of the machinery, but looks like the fix so far requires a ´few lines in a watch cache and listing from etcd using WithSort()

We can investigate both of those, to see if we can do them performantly.

Reply:
sorting (either in the apiserver from the watch cache, or in etcd for direct lists) would be in-memory... neither maintains an ordered index by revision:

https://github.com/kubernetes/kubernetes/blob/01e7b3040a1f47c613e6db85f9bccd727f5ef89c/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L441-L460

https://github.com/etcd-io/etcd/blob/24fc2a983ae39b4fe09cef3db434c812fdc97cff/etcdserver/apply.go#L313-L323

Reply:
opened exploratory PR with those changes at https://github.com/kubernetes/kubernetes/pull/74020

Reply:
cool, I've wrote up the issue here https://github.com/kubernetes/kubernetes/issues/74022 to direct the discussion and started writing some unit tests to prove the behavior locally, the PR is what I had in mind, thx for looking into it

Reply:
Until we sort out that approach in https://github.com/kubernetes/kubernetes/issues/74022 I've restricted using those values